base/shell/explorer/trayntfy.cpp

```diff
@@ -311,16 +373,86 @@ class CNotifyToolbar :
             }
         }
 
-        delete notifyItem;
+        HideBalloonTip(notifyItem);
+
         DeleteButton(index);
 
+        delete notifyItem;
+
+        return TRUE;
+    }
+
+    void UpdateBalloonTip(InternalIconData* notifyItem)
+    {
+        size_t len = 0;
+        if (SUCCEEDED(StringCchLength(notifyItem->szInfo, _countof(notifyItem->szInfo), &len)) && len > 0)
+        {
+            ShowBalloonTip(notifyItem);
+        }
+        else
+        {
+            HideBalloonTip(notifyItem);
+        }
+    }
+
+    static WPARAM GetTitleIcon(DWORD dwFlags, HICON hIcon)
+    {
+        if (dwFlags & NIIF_USER)
+            return reinterpret_cast<WPARAM>(hIcon);
+
+        return dwFlags & 3;
+    }
+
+    BOOL ShowBalloonTip(IN OUT InternalIconData *notifyItem)
+    {
+        DbgPrint("ShowBalloonTip called for flags=%x text=%ws; title=%ws", notifyItem->dwInfoFlags, notifyItem->szInfo, notifyItem->szInfoTitle);
+
+        // TODO: Queueing -> NIF_REALTIME? (Vista+)
+        // TODO: NIIF_NOSOUND, Vista+ flags
+        
+        const WPARAM icon = GetTitleIcon(notifyItem->dwInfoFlags, notifyItem->hIcon);
+        BOOL ret = m_Balloons->SetTitle(notifyItem->szInfoTitle, icon);
+        if (!ret)
+            DbgPrint("SetTitle failed, GetLastError=%d", GetLastError());
+
+        const int index = FindItemByIconData(notifyItem, NULL);
+        RECT rc;
+        GetItemRect(index, &rc);
+        ClientToScreen(&rc); // I have no idea why this is needed! >_<
```

Comment:
Because GetItemRect gets the rectangle relative to itself.

> Pointer to a RECT structure that receives **the client coordinates** of the bounding rectangle.

Reply:
I have no idea what you mean.

Reply:
GetItemRect receives the coordinates in _client_ coordinates.

Reply:
~~I know, that's the weird part: in the test app I wrote, the client coordinates work just fine...~~

I just looked myself, and noticed that I used GetWindowRect, not GetClientRect. LOL... No onder I was confused!